crates/storage/backend/fjall.rs

```diff
@@ -0,0 +1,175 @@
+    }
+
+    fn clear_table(&self, table: &'static str) -> Result<(), StoreError> {
+        let handle = self.get_partition(table)?;
+        self.keyspace.delete_partition(handle).unwrap();
+        init_partition(table, &self.keyspace)?;
+        Ok(())
+    }
+
+    fn begin_read(&self) -> Result<Box<dyn crate::api::StorageRoTx + '_>, StoreError> {
+        let backend = self.clone();
+        Ok(Box::new(FjallRoTx { backend }))
+    }
+
+    fn begin_write(&self) -> Result<Box<dyn crate::api::StorageRwTx + 'static>, StoreError> {
+        let backend = self.clone();
+        let write_batch = self.keyspace.batch();
+        let ro_tx = FjallRoTx { backend };
+        Ok(Box::new(FjallRwTx { ro_tx, write_batch }))
+    }
+
+    fn begin_locked(
+        &self,
+        table_name: &'static str,
+    ) -> Result<Box<dyn crate::api::StorageLocked>, StoreError> {
+        let snapshot = self.get_partition(table_name)?.snapshot();
+        Ok(Box::new(FjallLockedTx { snapshot }))
+    }
+}
+
+// Helper method to initialize a single partition
+fn init_partition(table_name: &str, keyspace: &Keyspace) -> Result<PartitionHandle, StoreError> {
+    let opts = PartitionCreateOptions::default().max_memtable_size(64 * 1024 * 1024);
+    let partition = keyspace.open_partition(table_name, opts).unwrap();
+    Ok(partition)
+}
+
+impl FjallBackend {
+    fn get_partition(&self, table_name: &str) -> Result<PartitionHandle, StoreError> {
+        // Use default opts, since partition should already exist
+        let opts = PartitionCreateOptions::default();
+        let partition = self.keyspace.open_partition(table_name, opts).unwrap();
+        Ok(partition)
+    }
+}
+
+struct FjallRoTx {
+    backend: FjallBackend,
+}
+
+impl StorageRoTx for FjallRoTx {
+    fn get(&self, table: &'static str, key: &[u8]) -> Result<Option<Vec<u8>>, StoreError> {
+        let partition = self.backend.get_partition(table).unwrap();
+        let value = partition.get(key).unwrap();
+        Ok(value.map(|v| v.to_vec()))
+    }
+
+    fn prefix_iterator(
+        &self,
+        table: &'static str,
+        prefix: &[u8],
+    ) -> Result<Box<dyn Iterator<Item = crate::api::PrefixResult> + '_>, StoreError> {
+        let partition = self.backend.get_partition(table).unwrap();
+        Ok(Box::new(partition.prefix(prefix).map(|res| {
+            let (k, v) = res.unwrap();
+            Ok((k.to_vec().into_boxed_slice(), v.to_vec().into_boxed_slice()))
+        })))
+    }
+}
+
+struct FjallRwTx {
+    ro_tx: FjallRoTx,
+    write_batch: WriteBatch,
+}
+
+impl StorageRoTx for FjallRwTx {
+    fn get(&self, table: &'static str, key: &[u8]) -> Result<Option<Vec<u8>>, StoreError> {
+        self.ro_tx.get(table, key)
+    }
+
+    fn prefix_iterator(
+        &self,
+        table: &'static str,
+        prefix: &[u8],
+    ) -> Result<Box<dyn Iterator<Item = crate::api::PrefixResult> + '_>, StoreError> {
+        self.ro_tx.prefix_iterator(table, prefix)
+    }
+}
```

Comment:
Question: does this support read-your-own-write? I think this skips the batch. I'm not sure it's a pattern we use.

Reply:
It indeed skips the batch. I just checked and we don't ever use that pattern. For now, I'm going to remove the requirement `StorageRoTx` in the `StorageRwTx` trait (in the original PR). We can reintroduce it if we deem it necessary.